core/services/headtracker/head_saver.go

```diff
@@ -11,137 +11,150 @@ import (
 		return nil
 	}
-	return ht.heads[0]
+	if ch.ChainLength() < hs.config.EvmFinalityDepth() {
+		hs.logger.Debugw("chain shorter than EvmFinalityDepth", "chainLen", ch.ChainLength(), "evmFinalityDepth", hs.config.EvmFinalityDepth())
+	}
+	return ch
 }
 
-func (ht *HeadSaver) Chain(hash common.Hash) *eth.Head {
-	ht.mu.RLock()
-	defer ht.mu.RUnlock()
+func (hs *headSaver) Chain(hash common.Hash) *eth.Head {
+	hs.mu.RLock()
+	defer hs.mu.RUnlock()
+
+	return hs.headByHash(hash)
+}
 
-	h := ht.headByHash(hash)
-	if h == nil {
+func (hs *headSaver) latestChain() *eth.Head {
+	if len(hs.heads) == 0 {
 		return nil
 	}
-	return h
+	return hs.heads[0]
 }
 
-// note: not thread-safe
-func (ht *HeadSaver) headByHash(hash common.Hash) (h *eth.Head) {
-	for _, h := range ht.heads {
+func (hs *headSaver) headByHash(hash common.Hash) (h *eth.Head) {
+	for _, h := range hs.heads {
 		if h.Hash == hash {
 			return h
 		}
 	}
 	return nil
 }
 
-// note: not thread-safe
-func (ht *HeadSaver) addHead(h *eth.Head, historyDepth int) {
-	ht.addHeads([]*eth.Head{h}, historyDepth)
+func (hs *headSaver) addHead(h *eth.Head, historyDepth uint) {
+	hs.addHeads([]*eth.Head{h}, historyDepth)
 }
 
-// note: not thread-safe
-func (ht *HeadSaver) addHeads(hs []*eth.Head, historyDepth int) {
-	hMap := make(map[common.Hash]*eth.Head, len(ht.heads)+len(hs))
-	for _, head := range append(ht.heads, hs...) {
+func (hs *headSaver) addHeads(newHeads []*eth.Head, historyDepth uint) {
+	headsMap := make(map[common.Hash]*eth.Head, len(hs.heads)+len(newHeads))
+	for _, head := range append(hs.heads, newHeads...) {
 		if head.Hash == head.ParentHash {
-			// shouldn't happen but its untrusted input
-			ht.logger.Errorf("ignoring head %s that points to itself", head)
+			// shouldn't happen but it is untrusted input
+			hs.logger.Errorf("ignoring head %s that points to itself", head)
 			continue
 		}
 		// copy all head objects to avoid races when a previous head chain is used
 		// elsewhere (since we mutate Parent here)
-		cphead := *head
-		cphead.Parent = nil // always build it from scratch in case it points to a head too old to be included
+		headCopy := *head
+		headCopy.Parent = nil // always build it from scratch in case it points to a head too old to be included
 		// map eliminates duplicates
-		hMap[head.Hash] = &cphead
+		headsMap[head.Hash] = &headCopy
 	}
-	heads := make([]*eth.Head, len(hMap))
+	heads := make([]*eth.Head, len(headsMap))
 	// unsorted unique heads
 	{
 		var i int
-		for _, head := range hMap {
+		for _, head := range headsMap {
 			heads[i] = head
 			i++
 		}
 	}
 	// sort the heads
 	sort.SliceStable(heads, func(i, j int) bool {
-		// sorting from highest number to lowest
+		// sorting from the highest number to lowest
 		return heads[i].Number > heads[j].Number
 	})
 	// cut off the oldest
-	if len(heads) > historyDepth {
+	if uint(len(heads)) > historyDepth {
 		heads = heads[:historyDepth]
 	}
 	// assign parents
-	for i := 0; i < len(heads)-1; i++ {
```

Comment:
was this off-by-one error or intentional? 
this does not seem affect tests...

Reply:
Perhaps there was an assumption that the last one would not have a parent, but it seems safe to do even if that is the case.

Maybe it would be simpler to switch to a `for _, head := range heads` now too?

Reply:
Fixed

Reply:
I remember writing this, and the logic was definitely intentional.

I can't remember why, but I suspect it will fail a test somewhere if you change it.

Reply:
Reverted to the original version.